[java] java.lang.ClassFormatError: Absent Code attribute in method that is not native or abstract in class file javax/faces/application/FacesMessage$Severity

**Affects PMD Version:** 6.3.0, 6.4.0-SNAPSHOT

**Rule:**: ---

**Description:**

My attempt to add `maven-pmd-plugin` to the Primefaces project failed with Maven 3.5.2 (and 3.5.3) due to

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.9.0:pmd (pmd) on project primefaces: Execution pmd of goal org.apache.maven.plugins:maven-pmd-plugin:3.9.0:pmd failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-pmd-plugin:3.9.0:pmd: java.lang.ClassFormatError: Absent Code attribute in method that is not native or abstract in class file javax/faces/application/FacesMessage$Severity
[ERROR] -----------------------------------------------------
[ERROR] realm =    plugin>org.apache.maven.plugins:maven-pmd-plugin:3.9.0
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
[ERROR] urls[0] = file:/home/travis/.m2/repository/org/apache/maven/plugins/maven-pmd-plugin/3.9.0/maven-pmd-plugin-3.9.0.jar
[ERROR] urls[1] = file:/home/travis/.m2/repository/net/sourceforge/pmd/pmd-core/6.4.0-SNAPSHOT/pmd-core-6.4.0-SNAPSHOT.jar
[ERROR] urls[2] = file:/home/travis/.m2/repository/com/beust/jcommander/1.48/jcommander-1.48.jar
[ERROR] urls[3] = file:/home/travis/.m2/repository/jaxen/jaxen/1.1.6/jaxen-1.1.6.jar
[ERROR] urls[4] = file:/home/travis/.m2/repository/net/java/dev/javacc/javacc/5.0/javacc-5.0.jar
[ERROR] urls[5] = file:/home/travis/.m2/repository/net/sourceforge/saxon/saxon/9.1.0.8/saxon-9.1.0.8.jar
[ERROR] urls[6] = file:/home/travis/.m2/repository/org/apache/commons/commons-lang3/3.7/commons-lang3-3.7.jar
[ERROR] urls[7] = file:/home/travis/.m2/repository/org/ow2/asm/asm/6.1.1/asm-6.1.1.jar
[ERROR] urls[8] = file:/home/travis/.m2/repository/com/google/code/gson/gson/2.5/gson-2.5.jar
[ERROR] urls[9] = file:/home/travis/.m2/repository/net/sourceforge/saxon/saxon/9.1.0.8/saxon-9.1.0.8-dom.jar
[ERROR] urls[10] = file:/home/travis/.m2/repository/net/sourceforge/pmd/pmd-java/6.4.0-SNAPSHOT/pmd-java-6.4.0-SNAPSHOT.jar
[ERROR] urls[11] = file:/home/travis/.m2/repository/net/sourceforge/pmd/pmd-javascript/6.4.0-SNAPSHOT/pmd-javascript-6.4.0-SNAPSHOT.jar
[ERROR] urls[12] = file:/home/travis/.m2/repository/org/mozilla/rhino/1.7.7.2/rhino-1.7.7.2.jar
[ERROR] urls[13] = file:/home/travis/.m2/repository/junit/junit/4.12/junit-4.12.jar
[ERROR] urls[14] = file:/home/travis/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar
[ERROR] urls[15] = file:/home/travis/.m2/repository/net/sourceforge/pmd/pmd-jsp/6.4.0-SNAPSHOT/pmd-jsp-6.4.0-SNAPSHOT.jar
[ERROR] urls[16] = file:/home/travis/.m2/repository/org/sonatype/sisu/sisu-inject-bean/1.4.2/sisu-inject-bean-1.4.2.jar
[ERROR] urls[17] = file:/home/travis/.m2/repository/org/sonatype/sisu/sisu-guice/2.1.7/sisu-guice-2.1.7-noaop.jar
[ERROR] urls[18] = file:/home/travis/.m2/repository/org/slf4j/jul-to-slf4j/1.7.25/jul-to-slf4j-1.7.25.jar
[ERROR] urls[19] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-sink-api/1.7/doxia-sink-api-1.7.jar
[ERROR] urls[20] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-logging-api/1.7/doxia-logging-api-1.7.jar
[ERROR] urls[21] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-decoration-model/1.7/doxia-decoration-model-1.7.jar
[ERROR] urls[22] = file:/home/travis/.m2/repository/org/codehaus/plexus/plexus-component-annotations/1.6/plexus-component-annotations-1.6.jar
[ERROR] urls[23] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-site-renderer/1.7/doxia-site-renderer-1.7.jar
[ERROR] urls[24] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-core/1.7/doxia-core-1.7.jar
[ERROR] urls[25] = file:/home/travis/.m2/repository/commons-lang/commons-lang/2.4/commons-lang-2.4.jar
[ERROR] urls[26] = file:/home/travis/.m2/repository/xmlunit/xmlunit/1.5/xmlunit-1.5.jar
[ERROR] urls[27] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-skin-model/1.7/doxia-skin-model-1.7.jar
[ERROR] urls[28] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-module-xhtml/1.7/doxia-module-xhtml-1.7.jar
[ERROR] urls[29] = file:/home/travis/.m2/repository/org/apache/maven/doxia/doxia-module-fml/1.7/doxia-module-fml-1.7.jar
[ERROR] urls[30] = file:/home/travis/.m2/repository/org/codehaus/plexus/plexus-i18n/1.0-beta-7/plexus-i18n-1.0-beta-7.jar
[ERROR] urls[31] = file:/home/travis/.m2/repository/org/codehaus/plexus/plexus-velocity/1.1.7/plexus-velocity-1.1.7.jar
[ERROR] urls[32] = file:/home/travis/.m2/repository/org/apache/velocity/velocity/1.7/velocity-1.7.jar
[ERROR] urls[33] = file:/home/travis/.m2/repository/org/apache/velocity/velocity-tools/2.0/velocity-tools-2.0.jar
[ERROR] urls[34] = file:/home/travis/.m2/repository/commons-beanutils/commons-beanutils/1.7.0/commons-beanutils-1.7.0.jar
[ERROR] urls[35] = file:/home/travis/.m2/repository/commons-digester/commons-digester/1.8/commons-digester-1.8.jar
[ERROR] urls[36] = file:/home/travis/.m2/repository/commons-chain/commons-chain/1.1/commons-chain-1.1.jar
[ERROR] urls[37] = file:/home/travis/.m2/repository/commons-validator/commons-validator/1.3.1/commons-validator-1.3.1.jar
[ERROR] urls[38] = file:/home/travis/.m2/repository/dom4j/dom4j/1.1/dom4j-1.1.jar
[ERROR] urls[39] = file:/home/travis/.m2/repository/oro/oro/2.0.8/oro-2.0.8.jar
[ERROR] urls[40] = file:/home/travis/.m2/repository/sslext/sslext/1.2-0/sslext-1.2-0.jar
[ERROR] urls[41] = file:/home/travis/.m2/repository/org/apache/struts/struts-core/1.3.8/struts-core-1.3.8.jar
[ERROR] urls[42] = file:/home/travis/.m2/repository/antlr/antlr/2.7.2/antlr-2.7.2.jar
[ERROR] urls[43] = file:/home/travis/.m2/repository/org/apache/struts/struts-taglib/1.3.8/struts-taglib-1.3.8.jar
[ERROR] urls[44] = file:/home/travis/.m2/repository/org/apache/struts/struts-tiles/1.3.8/struts-tiles-1.3.8.jar
[ERROR] urls[45] = file:/home/travis/.m2/repository/commons-collections/commons-collections/3.2.1/commons-collections-3.2.1.jar
[ERROR] urls[46] = file:/home/travis/.m2/repository/org/apache/maven/reporting/maven-reporting-api/3.0/maven-reporting-api-3.0.jar
[ERROR] urls[47] = file:/home/travis/.m2/repository/org/apache/maven/reporting/maven-reporting-impl/3.0.0/maven-reporting-impl-3.0.0.jar
[ERROR] urls[48] = file:/home/travis/.m2/repository/org/sonatype/aether/aether-util/1.7/aether-util-1.7.jar
[ERROR] urls[49] = file:/home/travis/.m2/repository/org/codehaus/plexus/plexus-interpolation/1.14/plexus-interpolation-1.14.jar
[ERROR] urls[50] = file:/home/travis/.m2/repository/org/sonatype/plexus/plexus-sec-dispatcher/1.3/plexus-sec-dispatcher-1.3.jar
[ERROR] urls[51] = file:/home/travis/.m2/repository/org/sonatype/plexus/plexus-cipher/1.4/plexus-cipher-1.4.jar
[ERROR] urls[52] = file:/home/travis/.m2/repository/org/apache/maven/shared/maven-shared-utils/3.2.0/maven-shared-utils-3.2.0.jar
[ERROR] urls[53] = file:/home/travis/.m2/repository/org/codehaus/plexus/plexus-resources/1.0.1/plexus-resources-1.0.1.jar
[ERROR] urls[54] = file:/home/travis/.m2/repository/org/codehaus/plexus/plexus-utils/3.0.24/plexus-utils-3.0.24.jar
[ERROR] urls[55] = file:/home/travis/.m2/repository/org/apache/httpcomponents/httpclient/4.5.3/httpclient-4.5.3.jar
[ERROR] urls[56] = file:/home/travis/.m2/repository/org/apache/httpcomponents/httpcore/4.4.6/httpcore-4.4.6.jar
[ERROR] urls[57] = file:/home/travis/.m2/repository/commons-logging/commons-logging/1.2/commons-logging-1.2.jar
[ERROR] urls[58] = file:/home/travis/.m2/repository/commons-codec/commons-codec/1.9/commons-codec-1.9.jar
[ERROR] urls[59] = file:/home/travis/.m2/repository/commons-io/commons-io/2.5/commons-io-2.5.jar
[ERROR] Number of foreign imports: 1
[ERROR] import: Entry[import  from realm ClassRealm[maven.api, parent: null]]
```

Please refer to [this build with 6.3.0](https://travis-ci.org/krichter722/primefaces/jobs/381918245) and [this build with 6.4.0-SNAPSHOT](https://travis-ci.org/krichter722/primefaces/builds/381923127) for examples.

This issue does not occur with Maven 3.3.9! It is similar to #328, but this separate report has been created because it has been requested there.

**Code Sample demonstrating the issue:** ---

**Running PMD through:** Maven (Plugin version 3.9.0)

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -43,6 +43,7 @@ This is a bug fixing release.
     *   [#1018](https://github.com/pmd/pmd/issues/1018): \[java] Performance degradation of 250% between 6.1.0 and 6.2.0
 *   java
     *   [#1077](https://github.com/pmd/pmd/issues/1077): \[java] Analyzing enum with lambda passed in constructor fails with "The enclosing scope must exist."
+    *   [#1131](https://github.com/pmd/pmd/issues/1131): \[java] java.lang.ClassFormatError: Absent Code attribute in method that is not native or abstract in class file javax/faces/application/FacesMessage$Severity
 *   java-bestpractices
     *   [#527](https://github.com/pmd/pmd/issues/572): \[java] False Alarm of JUnit4TestShouldUseTestAnnotation on Predicates
     *   [#1063](https://github.com/pmd/pmd/issues/1063): \[java] MissingOverride is triggered in illegal places

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java
@@ -1249,7 +1249,7 @@ public class ClassTypeResolver extends JavaParserVisitorAdapter {
                     + qualifiedName.substring(qualifiedName.lastIndexOf('.') + 1);
             try {
                 myType = pmdClassLoader.loadClass(qualifiedNameInner);
-            } catch (Exception ignored) {
+            } catch (ClassNotFoundException ignored) {
                 // ignored, we'll try again with a different package name/fqcn
             } catch (LinkageError e) {
                 // we found the class, but there is a problem with it (see https://github.com/pmd/pmd/issues/1131)

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java
@@ -1234,7 +1234,12 @@ public class ClassTypeResolver extends JavaParserVisitorAdapter {
                 } catch (ClassNotFoundException e) {
                     myType = processOnDemand(qualifiedName);
                 } catch (LinkageError e) {
-                    myType = processOnDemand(qualifiedName);
+                    // we found the class, but there is a problem with it (see https://github.com/pmd/pmd/issues/1131)
+                    if (LOG.isLoggable(Level.FINE)) {
+                        LOG.log(Level.FINE, "Tried to load class " + qualifiedName + " from on demand import, "
+                                + "with an incomplete classpath.", e);
+                    }
+                    return;
                 }
             }
         }
@@ -1246,6 +1251,13 @@ public class ClassTypeResolver extends JavaParserVisitorAdapter {
                 myType = pmdClassLoader.loadClass(qualifiedNameInner);
             } catch (Exception ignored) {
                 // ignored, we'll try again with a different package name/fqcn
+            } catch (LinkageError e) {
+                // we found the class, but there is a problem with it (see https://github.com/pmd/pmd/issues/1131)
+                if (LOG.isLoggable(Level.FINE)) {
+                    LOG.log(Level.FINE, "Tried to load class " + qualifiedNameInner + " from on demand import, "
+                            + "with an incomplete classpath.", e);
+                }
+                return;
             }
         }
         if (myType == null && qualifiedName != null && !qualifiedName.contains(".")) {
@@ -1302,8 +1314,11 @@ public class ClassTypeResolver extends JavaParserVisitorAdapter {
         try {
             pmdClassLoader.loadClass(fullyQualifiedClassName);
             return true; // Class found
-        } catch (ClassNotFoundException | NoClassDefFoundError e) {
+        } catch (ClassNotFoundException e) {
             return false;
+        } catch (LinkageError e2) {
+            // Class exists, but may be invalid (see https://github.com/pmd/pmd/issues/1131)
+            return true;
         }
     }
 
@@ -1311,6 +1326,12 @@ public class ClassTypeResolver extends JavaParserVisitorAdapter {
         try {
             return pmdClassLoader.loadClass(fullyQualifiedClassName);
         } catch (ClassNotFoundException e) {
+            return null;
+        } catch (LinkageError e2) {
+            if (LOG.isLoggable(Level.FINE)) {
+                LOG.log(Level.FINE, "Tried to load class " + fullyQualifiedClassName + " from on demand import, "
+                        + "with an incomplete classpath.", e2);
+            }
             return null;
         }
     }

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/TypeHelper.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/TypeHelper.java
@@ -42,6 +42,9 @@ public final class TypeHelper {
                 // The requested type is not on the auxclasspath. This might happen, if the type node
                 // is probed for a specific type (e.g. is is a JUnit5 Test Annotation class).
                 // Failing to resolve clazzName does not necessarily indicate an incomplete auxclasspath.
+            } catch (final LinkageError expected) {
+                // We found the class but it's invalid / incomplete. This may be an incomplete auxclasspath
+                // if it was a NoClassDefFoundError. TODO : Report it?
             }
         }
 

